### PR TITLE
Adding a BulkMutation throttler.

### PR DESCRIPTION
**Pull request by sduskis**
Adding a throttler and throttling options.

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BulkOptions.java [hunk 1]
```diff
@@ -50,13 +50,25 @@
...
+   */
+  public static final int BIGTABLE_BULK_THROTTLE_TARGET_MS_DEFAULT = 40;
+
+  /**
    * The maximum amount of time a row will be buffered for. By default 0: indefinitely.
    */
   public static long BIGTABLE_BULK_AUTOFLUSH_MS_DEFAULT = 0;
 
   // Default rpc count per channel.
   /** Constant <code>BIGTABLE_MAX_INFLIGHT_RPCS_PER_CHANNEL_DEFAULT=50</code> */
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BulkOptions.java]: Comment is out of sync

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java [hunk 2]
```diff
@@ -64,7 +64,7 @@ private static int getDefaultDataChannelCount() {
     // 10 Channels seemed to work well on a 4 CPU machine, and this seems to scale well for higher
     // CPU machines. Use no more than 250 Channels by default.
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java]: Comment out of sync.

Why the change to 5?

File: bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java [hunk 3]
```diff
@@ -216,6 +216,19 @@
   public static final String BIGTABLE_BUFFERED_MUTATOR_MAX_MEMORY_KEY =
       "google.bigtable.buffered.mutator.max.memory";
 
+  /**
+   * Turn on a feature that will reduce the likelihood of BufferedMutator overloading a Cloud
+   * Bigtable cluster.
+   */
+  public static final String BIGTABLE_BUFFERED_MUTATOR_ENABLE_THROTTLING =
+      "google.bigtable.buffered.mutator.enable.throttling";
```
Comment (garye) [inline bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java]: nit - both options should maybe have the same prefix: google.bigtable.buffered.mutator.throttling

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/ResourceLimiter.java [hunk 4]
```diff
@@ -89,6 +91,22 @@ public int getMaxInFlightRpcs() {
   }
 
   /**
+   * <p>Getter for the field <code>currentInFlightRpcs</code>.</p>
+   *
+   * @return The current number of allowed in-flight RPCs
+   */
+  public int getCurrentInFlightRpcs() {
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/ResourceLimiter.java]: The name of the method doesn't seem to go with the comment. Is it allowed RPCs or currently executing RPCs?

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.concurrent.TimeUnit;
+
+import com.google.cloud.bigtable.grpc.BigtableSessionSharedThreadPools;
+
+public class BulkMutationThrottling {
+
+  private static boolean isThrottling = false;
+
+  public static synchronized void throttle(final ResourceLimiter resourceLimiter,
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java]: There are a lot of heuristics and magic here, would be nice if there was some explanation as to why things are done the way they are (eventually - maybe once this is no longer an experimental feature).

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      return;
+    }
+    final long highTargetNanos =
+        TimeUnit.MILLISECONDS.toNanos((long) (bulkMutationRpcTargetMs * 1.2));
+    final long lowTargetNanos =
+        TimeUnit.MILLISECONDS.toNanos((long) (bulkMutationRpcTargetMs * 0.8));
+    // Change at 5% increments
+    final int throttlingChangeStep = resourceLimiter.getMaxInFlightRpcs() / 20;
+    int current = resourceLimiter.getCurrentInFlightRpcs();
+    resourceLimiter.setCurrentInFlightRpcs(current / 2);
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java]: Why does enabling throttling reduce the # of outstanding RPCs in half right away, even before latency measurements are examined?

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java (hunk 2 above)
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java]: Our internal performance tests showed that we should be adding fewer requests on the same channel.  I doubled the channels, and significantly reduced the outstanding RPCs.  I think that this deserves a separate PR, so I'll undo these changes.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ */
+package com.google.cloud.bigtable.grpc.async;
+
+import java.util.concurrent.TimeUnit;
+
+import com.google.cloud.bigtable.grpc.BigtableSessionSharedThreadPools;
+
+public class BulkMutationThrottling {
+
+  private static boolean isThrottling = false;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java]: What happens if a user creates 2 sessions?

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java (hunk 5 above)
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationThrottling.java]: I don't love this being a static method, I feel like its something that the session should own

**Comment by sduskis** [comment]
@igorbernstein2 I moved the throttling logic to `ResourceLimiter` so that there isn't a separate static method.  That said, `ResourceLimiter` is accessed as a static variable on `BigtableSession`.  In dataflow, `BigtableSession`s are ephemeral.  Each bundle gets its own `BigtableSession`.  There are a set of static variables (`cachedDataChannelPool` and `resourceLimiter` in `BigtableSession`, a couple of executors in grpc, and `BigtableSessionSharedThreadPools`) that allow for performance and VM level controls, like the throttling mechanism.

**Comment by sduskis** [comment]
I added quite a few comments, and moved the throttling logic to ResourceLimiter.  PTAL.

**Comment by igorbernstein2** [comment]
Looks good, one last request: can you add a dropwizard gauge for currentInFlightMaxRpcs...it should make it easier for users to tune & debug performance

**Comment by sduskis** [comment]
We discussed metrics offline.  The scope of the work is large enough to require another PR.

**Comment by sduskis** [comment]
@garye, I'm going to merge this.  If we need to clarify docs, let's please discuss this off line.
